find_smallest_cycle: Optimize gather_deps to abort early

zmedico opened the pull request:
If gather_deps is traversing a cycle that is greater than
or equal to the size of the current smallest_cycle, then
abort early. Also abort early if we traverse a node
encountered in a previous cycle for the same ignore_priority,
since that means the two cycles are identical.

On my laptop, this brings the `emerge -pe @world` time down
to 3m28.884s, compared to 10m44.268s with portage-3.0.55.

Bug: https://bugs.gentoo.org/917660

thesamesam commented on lib/_emerge/depgraph.py:
```diff
@@ -9172,9 +9174,17 @@ def gather_deps(ignore_priority, mergeable_nodes, selected_nodes, node):
                 # (these occur when new portage blocks an older package version).
                 return False
             selected_nodes.add(node)
+            if smallest_cycle is not None and len(selected_nodes) >= len(
```
I've not read over this yet outside of the gh viewer so it's harder to see the surrounding stuff - but I wonder if we should do something like store a tuple of `(ignore_priority, size)` or maybe even `(greatest_priority_ignored, size)` and go after that instead. (It might be that this isn't relevant because we don't take the path in this case, so ignore me if so)

zmedico replied:
I'm not sure how the ignore_priority part of this tuple is useful. The ignore_priority objects are not really comparable to each other, so I'm not sure what you would want to do with it.

thesamesam replied:
The idea would be that if we had `ignore_priority` set for a cycle solution, we'd prefer one unset unless we've tried for a while and can't get anything better?

thesamesam replied:
Maybe I'm trying to reinvent more fine-grained priority stuff actually and we shouldn't worry here.

zmedico replied:
It starts with local_priority_range.MEDIUM_POST and only increases ignore_priority from there. When find_smallest_cycle is called, we assume that we need to ignore something, since otherwise it would have selected something earlier here:
```python
                for i in range(priority_range.NONE, priority_range.MEDIUM_SOFT + 1):
                    ignore_priority = priority_range.ignore_priority[i]
                    nodes = get_nodes(ignore_priority=ignore_priority)
```

thesamesam replied:
Got it, thank you!

thesamesam commented:
Performance still seems poor with this, unfortunately.

I took a few stacktraces:
```
Calculating dependencies \--Return--
> /usr/lib/python-exec/python3.12/emerge(28)debug_signal()->None
-> pdb.set_trace()
(Pdb) bt
  /usr/lib/python-exec/python3.12/emerge(88)<module>()
-> main()
  /usr/lib/python-exec/python3.12/emerge(57)main()
-> retval = emerge_main()
  /usr/lib/python3.12/site-packages/_emerge/main.py(1308)emerge_main()
-> return run_action(emerge_config)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(4013)run_action()
-> retval = action_build(emerge_config, spinner=spinner)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(409)action_build()
-> success, mydepgraph, favorites = backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11631)backtrack_depgraph()
-> success, mydepgraph, favorites, backtracked, max_retries = _backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11685)_backtrack_depgraph()
-> success, favorites = mydepgraph.select_files(myfiles)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(4686)select_files()
-> return self._select_files(args)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5120)_select_files()
-> return self._resolve(myfavorites)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5323)_resolve()
-> self.altlist()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(8865)altlist()
-> ) = self._serialize_tasks()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9395)_serialize_tasks()                                                                                                                                 -> cycle_digraph = mygraph.copy()                                                                                                                                                                                 /usr/lib/python3.12/site-packages/portage/util/digraph.py(279)clone()                                                                                                                                         -> priorities_clone = memo.get(id(priorities))                                                                                                                                                                  > /usr/lib/python-exec/python3.12/emerge(28)debug_signal()->None
-> pdb.set_trace()
(Pdb) c
[...]
(Pdb) bt
  /usr/lib/python-exec/python3.12/emerge(88)<module>()
-> main()
  /usr/lib/python-exec/python3.12/emerge(57)main()
-> retval = emerge_main()
  /usr/lib/python3.12/site-packages/_emerge/main.py(1308)emerge_main()
-> return run_action(emerge_config)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(4013)run_action()
-> retval = action_build(emerge_config, spinner=spinner)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(409)action_build()
-> success, mydepgraph, favorites = backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11631)backtrack_depgraph()
-> success, mydepgraph, favorites, backtracked, max_retries = _backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11685)_backtrack_depgraph()
-> success, favorites = mydepgraph.select_files(myfiles)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(4686)select_files()
-> return self._select_files(args)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5120)_select_files()
-> return self._resolve(myfavorites)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5323)_resolve()
-> self.altlist()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(8865)altlist()
-> ) = self._serialize_tasks()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9382)_serialize_tasks()
-> selected_nodes, ignore_priority = find_smallest_cycle(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9349)find_smallest_cycle()
-> if gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(                                                                                                                                                                                            /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()                                                                                                                                      -> if not gather_deps(                                                                                                                                                                                            /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()                                                                                                                                      -> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
[...]
-> main()                                                                                                                                                                                            [2287/2408]
  /usr/lib/python-exec/python3.12/emerge(57)main()
-> retval = emerge_main()
  /usr/lib/python3.12/site-packages/_emerge/main.py(1308)emerge_main()
-> return run_action(emerge_config)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(4013)run_action()
-> retval = action_build(emerge_config, spinner=spinner)
  /usr/lib/python3.12/site-packages/_emerge/actions.py(409)action_build()
-> success, mydepgraph, favorites = backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11631)backtrack_depgraph()
-> success, mydepgraph, favorites, backtracked, max_retries = _backtrack_depgraph(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(11685)_backtrack_depgraph()
-> success, favorites = mydepgraph.select_files(myfiles)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(4686)select_files()
-> return self._select_files(args)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5120)_select_files()
-> return self._resolve(myfavorites)
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(5323)_resolve()
-> self.altlist()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(8865)altlist()
-> ) = self._serialize_tasks()
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9382)_serialize_tasks()
-> selected_nodes, ignore_priority = find_smallest_cycle(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9349)find_smallest_cycle()
-> if gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9182)gather_deps()
-> if not gather_deps(
  /usr/lib/python3.12/site-packages/_emerge/depgraph.py(9164)gather_deps()
-> if node not in mergeable_nodes:
  /usr/lib/python3.12/site-packages/_emerge/Task.py(24)__hash__()
-> def __hash__(self):
> /usr/lib/python-exec/python3.12/emerge(28)debug_signal()->None
-> pdb.set_trace()
(Pdb) c
```

thesamesam commented:
It's unhappy with the latest version too. Should I grab the same sort of poor-man's sampled stacktraces, profile properly, or some third thing?

zmedico commented:
> It's unhappy with the latest version too. Should I grab the same sort of poor-man's sampled stacktraces, profile properly, or some third thing?

Try again now, since I added a traversed_nodes set to prevent it from traversing the same cycle multiple times.

I also added a target_size parameter to find_smallest_cycle that currently defaults to 3, since a target_size of 1 drives up ignore_priority and ignores more dependencies than necessary.

thesamesam commented:
Thanks! It now takes about ~50-60s so it completes in reasonableish time, but quite a while longer than before. I wonder why my case is so pathological..

thesamesam commented:
I seem to be able to hit it on another machine with a similar world file and `/etc/portage`. It's a test environment so I should be able to give you access to it if needed.

zmedico commented:
> Thanks! It now takes about ~50-60s so it completes in reasonableish time, but quite a while longer than before (~10s). I wonder why my case is so pathological..

In order to reduce the time more, we need to avoid as many gather_deps calls as possible. There are cases where find_smallest_cycle can needlessly loop through all the priorities without finding any mergeable cycles. We can eliminate some nodes earlier without so may needless loops if we test them at the highest ignore_priority and and find that they are not part of a mergeable cycle even with the highest ignore_priority.

thesamesam commented:
Looks like it doesn't help unfortunately.

zmedico commented:
> Looks like it doesn't help unfortunately.

We'll have to revert 9206d5a75ecd2d9ae0fe63e57d28aa8061b5927e, since apparently the `len(smallest_cycle) == 1` loop termination condition is not optimal and it's too expensive. I'll add the revert to this MR and rebase on that, since the optimization to abort gather_deps early will still apply.

thesamesam commented:
Thanks Zac, lgtm

zmedico commented:
Revert is in cdc781349337fa755bc15773e2a87e4b41f5ff1e. Opened https://github.com/gentoo/portage/pull/1191 for the optimization.
